zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -735,10 +735,18 @@ private String getParentPathAndValidate(String path) throws BadArgumentsExceptio
     }
 
     private static int checkAndIncVersion(int currentVersion, int expectedVersion, String path) throws KeeperException.BadVersionException {
-        if (expectedVersion != -1 && expectedVersion != currentVersion) {
-            throw new KeeperException.BadVersionException(path);
+       if (expectedVersion != -1) {
+            if (expectedVersion != currentVersion) {
+                throw new KeeperException.BadVersionException(path);
+            }
+            if (expectedVersion == -2) {
+                return 0;
+            } else {
+                return currentVersion + 1;
+            }
+        } else {
+            return currentVersion + 1;
```

Comment:
The next version should be `0` when `currentVersion` is `-2` irrespective of `expectedVersion.

Reply:
Hey. @kezhuw Thanks for your suggestion. Can you explain it further? The current approach is to be as consistent as possible with the existing semantics of zookeeper, that is, except expectedVersion which is -1, all other values need to be checked for equality. 
So my question is why should we skip this check when it is -2?

Reply:
Thanks for your detailed explanation. I almost understand the background.
> The next version should be 0 when currentVersion is -2 irrespective of expectedVersion.

I'm not sure if my understanding of this sentence is correct, what we can agree on is that the currentVersion needs to be changed from -2 to 0.
But if currentVersion=-2, the equality check is skipped, I am worried that there may be safety risks?
For example, in the following scenario, two clients concurrently initiate setData requests. Counted as C1 and C2 respectively, the version numbers queried before C1 and C2 initiated the update were both -6, and then simultaneously initiated requests to ZkServer. The request initiated by C2 was successfully processed. C1 was delayed for several seconds due to network problems. Here, During this period, C2 initiated three more requests, causing the version number to become -2. Then the network between C1 and ZkServer returned to normal. At this time, ZkServer did not perform an equality check because currentVersion=-2, resulting in C1's request being successfully processed.
Is this situation in line with expectations? My personal understanding is that this is not in line with expectations for C2 and violates the semantics of CAS, because C1 successfully updated with an older version.

Reply:
I think below is enough for us to go.

```
-        return currentVersion + 1;
+        return currentVersion == -2 ? 0 : currentVersion + 1;
```

Reply:
```java
if (expectedVersion != -1) {
            if (expectedVersion != currentVersion) {
                throw new KeeperException.BadVersionException(path);
            }
            if (expectedVersion == -2) {
                return 0; // <-- this branch is unreachable
```

I suppose this is what @kezhuw means?

Reply:
@kezhuw Perhaps we can directly align the proposed `checkAndIncVersion`?

I'd suggest:

```java
    private static int checkAndIncVersion(int currentVersion, int expectedVersion, String path) throws KeeperException.BadVersionException {
        if (expectedVersion != -1 && expectedVersion != currentVersion) {
            throw new KeeperException.BadVersionException(path);
        }

        // Skip increase the data version to -1 by jumping from -2 to 0. Otherwise, the client will
        // receive a new data version -1. Now if the client wants to check the data version, it can
        // only pass -1 as the next expected version, but -1 as the expected version means do not
        // check the data version. So the client is unable to express the expected manner.
        return currentVersion != -2 ? currentVersion + 1 : 0;
    }
```

Reply:
In this way, we can avoid mentioning `-2` as a new magic number. `-2` is special because `-1` is special - let's talk about -1 instead converged.

Reply:
makes sense. This suggest version is more readable, concise and elegant